File: 4_grpc/random_server.py
from concurrent import futures
import logging
import random

import grpc
import random_pb2
import random_pb2_grpc

class RandomServer( random_pb2_grpc.RandomServiceServicer ):
    def __init__(self):
        self.min = 0
        self.max = 10

    def SetBounds( self, request, context ):
        self.min = request.min
        self.max = request.max
        return random_pb2.SetBoundsResponse()

    def NextRandom( self, request, context ):
        random_response = int(random.random() * (self.max - self.min + 1)) + self.min
        return random_pb2.NextRandomResponse(random=random_response)

def serve():
    server = grpc.server( futures.ThreadPoolExecutor( max_workers=10 ))
    random_pb2_grpc.add_RandomServiceServicer_to_server( RandomServer(), server )
    server.add_insecure_port( '[::]:50051' )
    server.start()
    print( "server ready" )
    server.wait_for_termination()


if __name__ == '__main__':
    logging.basicConfig()
    serve()

File: 4_grpc/random_server.cc

#include <iostream>
#include <string>
#include <memory>

#include <cstdlib>

#include <grpc++/grpc++.h>

#include "random.pb.h"
#include "random.grpc.pb.h"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::Status;

namespace if3040::grpc {

class RandomServer final : public RandomService::Service {
public:
    RandomServer() {}
    
    Status SetBounds( ServerContext * context,
                      const SetBoundsRequest *  request,
                      SetBoundsResponse  * response ) {
        min = request->min();
        max = request->max();
        return Status::OK;
    }
    
    Status NextRandom( ServerContext * context,
                       const NextRandomRequest * request,
                       NextRandomResponse * response ) {
        int random = rand() % (max - min + 1) + min;
        response->set_random(random);
        return Status::OK;
    }

private:
    int min = 0;
    int max = 10;
};

}


int main() {
    if3040::grpc::RandomServer random_server;
    
    std::string server_address( "0.0.0.0:50051" );
    ServerBuilder builder;
    builder.AddListeningPort( server_address, grpc_impl::InsecureServerCredentials() );
    builder.RegisterService( &random_server );
    std::unique_ptr< Server > server( builder.BuildAndStart() );
    std::cout << "Server listening on " << server_address << std::endl;
    server->Wait();
    
    return 0;
}
